importer/src/path_canonicalizer.rs

```diff
@@ -0,0 +1,378 @@
+                    }
+                    // modules expose symbols
+                    Symbol::Module(Module::Local(module)) => module
+                        .symbol_definitions()
+                        .find_map(|SymbolDefinition { name, symbol }| {
+                            (name == member).then_some(symbol.clone())
+                        })
+                        .ok_or_else(|| format!("symbol not found in `{location}`: `{member}`"))
+                        .and_then(|symbol| {
+                            match symbol {
+                                Symbol::Import(p) => {
+                                    // if we found an import, check it and continue from there
+                                    check_path(location, p.path, state)
+                                }
+                                symbol => {
+                                    // if we found any other symbol,
+                                    Ok((state, location.join(member.clone()), symbol))
+                                }
+                            }
+                        }),
+                    // external modules must have been turned into local ones before
+                    Symbol::Module(Module::External(_)) => unreachable!(),
+                    Symbol::Import(p) => {
+                        location.pop_back().unwrap();
+
+                        // redirect to `p`
+                        check_path(location, p.path.join(member.clone()), state)
+                    }
+                }
+            },
+        )
+        .map(|(mut state, absolute_path, symbol)| {
+            state
+                .paths
+                .insert((location, imported), absolute_path.clone());
+            (state, absolute_path, symbol)
+        })
+}
+
+/// Checks an import
+///
+/// # Errors
+///
+/// This function will return an error if the imported path does not resolve to anything
+fn check_import<T: Clone>(
+    // the location at which the import is made
+    location: AbsoluteSymbolPath,
+    // the imported path, relative to the location
+    imported: Import,
+    // the current state
+    state: State<'_, T>,
+) -> Result<State<'_, T>, String> {
+    let (state, _, _) = check_path(location, imported.path, state)?;
+
+    Ok(state)
+}
+
+fn generate_path_map<T: FieldElement>(program: &ASMProgram<T>) -> Result<PathMap, String> {
+    check_module(
+        // the location of the main module
+        AbsoluteSymbolPath::default(),
+        &program.main,
+        // an empty state starting from this module
+        State {
+            root: &program.main,
+            paths: Default::default(),
+        },
+    )
+    .map(|state| state.paths)
+}
+
+/// Checks a module
+///
+/// # Errors
+///
+/// This function will return an error if a name is not unique, or if any path in this module does not resolve to anything
+fn check_module<'a, T: Clone>(
+    location: AbsoluteSymbolPath,
+    module: &ASMModule<T>,
```

Comment:
Can we replace this by a trait so it can be used more generally?

Reply:
What do you mean by a trait? Like a visitor for the parse tree?